### PR TITLE
`r\shared_image`: Add support for `support_accelerated_network`

### LINKED ISSUE #15441: r/shared_image: support for advanced_networking_enabled
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

In December 2021 Microsoft added support for specifying features for image definitions.  Documentation for what can be supplied is in the documentation for the [az cli](https://github.com/MicrosoftDocs/azure-docs-cli/blob/main/latest/docs-ref-autogen/sig/image-definition.yml).

The issue is that when adding a new image version, if the source image has a feature configured, like IsAcceleratedNetworkSupported=true, and the definition does not have this feature, image creation fails.

### New or Affected Resource(s)

* azurerm_shared_image

### Potential Terraform Configuration

```hcl
resource "azurerm_shared_image" "image" {
  name                = "myImage"
  gallery_name        = "myGallery"
  resource_group_name = "myResourceGroup"
  location            = "centralus"
  os_type             = "linux"
  
  features {
    IsAcceleratedNetworkSupported = true
  }

  identifier {
    publisher = "myPublisher"
    offer     = "myOffer"
    sku       = "mysku"
  }

  purchase_plan {
    publisher = "myPublisher"
    product   = "myOffer"
    publisher = "mysku"
  }

}
```

### References

https://github.com/MicrosoftDocs/azure-docs-cli/blob/main/latest/docs-ref-autogen/sig/image-definition.yml
https://github.com/Azure/azure-powershell/pull/16621/files
https://docs.microsoft.com/en-us/powershell/module/az.compute/new-azgalleryimagedefinition?view=azps-7.2.0#examples -Example 4
https://docs.microsoft.com/en-us/rest/api/compute/gallery-images/update

### PR DESCRIPTION
**Pull request by myc2h6o**
Fix #15441
- Add new gallery image feature `support_accelerated_network`

### CODE REVIEW

**Comment by myc2h6o** [comment]
Test result:
$ TF_ACC=1 go test -v ./internal/services/compute -run='TestAccSharedImage_withSupportAcceleratedNetwork' -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSharedImage_withSupportAcceleratedNetwork
=== PAUSE TestAccSharedImage_withSupportAcceleratedNetwork
=== CONT  TestAccSharedImage_withSupportAcceleratedNetwork
--- PASS: TestAccSharedImage_withSupportAcceleratedNetwork (356.45s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/compute       357.203s

File: internal/services/compute/shared_image_resource.go [hunk 1]
```diff
@@ -187,14 +194,6 @@ func resourceSharedImageCreateUpdate(d *pluginsdk.ResourceData, meta interface{}
 		}
 	}
 
-	var features []compute.GalleryImageFeature
-	if d.Get("trusted_launch_enabled").(bool) {
-		features = append(features, compute.GalleryImageFeature{
-			Name:  utils.String("SecurityType"),
-			Value: utils.String("TrustedLaunch"),
-		})
-	}
```
Comment (tombuildsstuff) [inline internal/services/compute/shared_image_resource.go:196]: can we remove the expand function here and leave this as it was (adding the new field) - passing values out of `d` in functions makes it more likely we'll introduce bugs when refactoring etc

File: internal/services/compute/shared_image_resource.go [hunk 2]
```diff
@@ -438,3 +429,48 @@ func flattenGalleryImagePurchasePlan(input *compute.ImagePurchasePlan) []interfa
...
+	var features []compute.GalleryImageFeature
+
+	if d.Get("trusted_launch_enabled").(bool) {
+		features = append(features, compute.GalleryImageFeature{
+			Name:  utils.String("SecurityType"),
+			Value: utils.String("TrustedLaunch"),
+		})
+	}
+
+	if d.Get("support_accelerated_network").(bool) {
```
Comment (tombuildsstuff) [inline internal/services/compute/shared_image_resource.go:443]: (as above) can we move this into the Create function?

File: internal/services/compute/shared_image_resource.go [hunk 3]
```diff
@@ -438,3 +429,48 @@ func flattenGalleryImagePurchasePlan(input *compute.ImagePurchasePlan) []interfa
...
+		features = append(features, compute.GalleryImageFeature{
+			Name:  utils.String("IsAcceleratedNetworkSupported"),
+			Value: utils.String("true"),
+		})
+	}
+
+	return &features
+}
+
+func flattenAndSetGalleryImageFeatures(d *pluginsdk.ResourceData, features *[]compute.GalleryImageFeature) {
```
Comment (tombuildsstuff) [inline internal/services/compute/shared_image_resource.go:453]: (as above) can we move this into the read function - `AndSet` functions make it easier for bugs to be introduced

File: internal/services/compute/shared_image_resource.go [hunk 4]
```diff
@@ -438,3 +429,48 @@ func flattenGalleryImagePurchasePlan(input *compute.ImagePurchasePlan) []interfa
...
+			}
+		}
+
+		if featuresMap["SecurityType"] == "TrustedLaunch" {
+			trustedLaunchEnabled = true
+		}
+
+		if strings.EqualFold(featuresMap["IsAcceleratedNetworkSupported"], "true") {
+			supportAcceleratedNetwork = true
+		}
```
Comment (tombuildsstuff) [inline internal/services/compute/shared_image_resource.go:471]:
```suggestion
		for _, feature := range *features {
			if feature.Name == nil || feature.Value == nil {
				continue
			}
			
			if strings.EqualFold(*feature.Name, "SecurityType") {
				trustedLaunchEnabled = strings.EqualFold(*feature.Value, "TrustedLaunch")
			}
			
			if strings.EqualFold(*feature.Name, "IsAcceleratedNetworkSupported") {
				supportAcceleratedNetwork = strings.EqualFold(*feature.Value, "true")
			}
		}
```

File: internal/services/compute/shared_image_resource.go [hunk 5]
```diff
@@ -160,6 +161,12 @@ func resourceSharedImage() *pluginsdk.Resource {
 				ForceNew: true,
 			},
 
+			"support_accelerated_network": {
```
Comment (tombuildsstuff) [inline internal/services/compute/shared_image_resource.go:164]: can we document this field?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @myc2h6o 

Thanks for this PR.

I've taken a look through and left some comments inline, but if we can address those then we should be able to take another look.

Thanks!

File: internal/services/compute/shared_image_resource.go (hunk 5 above)
Comment (myc2h6o) [reply internal/services/compute/shared_image_resource.go:164]: oops, forgot the doc. Added

File: internal/services/compute/shared_image_resource.go (hunk 1 above)
Comment (myc2h6o) [reply internal/services/compute/shared_image_resource.go]: Got it, moved to Create.

File: internal/services/compute/shared_image_resource.go (hunk 2 above)
Comment (myc2h6o) [reply internal/services/compute/shared_image_resource.go]: Done

File: internal/services/compute/shared_image_resource.go (hunk 3 above)
Comment (myc2h6o) [reply internal/services/compute/shared_image_resource.go]: Done

File: internal/services/compute/shared_image_resource.go (hunk 4 above)
Comment (myc2h6o) [reply internal/services/compute/shared_image_resource.go]: Updated. Looks more straightforward since we will probably not having too many features here

**Comment by myc2h6o** [comment]
Hi @tombuildsstuff I've updated the pr, please take a look

**Comment by myc2h6o** [comment]
New test result after change:
image

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/shared_image.html.markdown [hunk 6]
```diff
@@ -82,6 +82,8 @@ The following arguments are supported:
 
 * `trusted_launch_enabled` - (Optional) Specifies if Trusted Launch has to be enabled for the Virtual Machine created from the Shared Image. Defaults to `false`. Changing this forces a new resource to be created.
 
+* `support_accelerated_network` - (Optional) Specifies if the Shared Image supports Accelerated Network. Defaults to `false`. Changing this forces a new resource to be created.
```
[the remark this answers is not in the document]
Comment (myc2h6o) [reply website/docs/r/shared_image.html.markdown]: Done.

**Comment by myc2h6o** [comment]
Test result after update
image

**Comment by myc2h6o** [comment]
Hi @katbyte I've updated the field name, could you take another look?

**Review by katbyte — APPROVED** [review]
Thanks @myc2h6o - LGTM 🏗️

**Pull request merged**